### ISSUE TITLE
Warn against empy build methods

### ISSUE THREAD

**Issue by rrousselGit** [rrousselGit/riverpod#3031]
Some folks do:

```dart
State? build() => null
```

This likely stems from a misusage of the Riverpod API and should therefore be discouraged

**Comment by rrousselGit** [comment]
This probably should happen only if `build` returns a Future/Stream

**Comment by xleon** [comment]
Hi, new to Riverpod here. This is my case: 
```dart
@Riverpod(keepAlive: true)
class Package extends _$Package {
  @override
  PackageInfo? build() {
    return null;
  }

  Future<void> init() async => state = await PackageInfo.fromPlatform();
}
```
Why?
Because I want to directly access the value rather than working with AsyncValue. 
Do you know a better way?

**Comment by xleon** [comment]
Ha!, found the answer to my question right in the docs: 
https://riverpod.dev/docs/concepts/scopes#initialization-of-synchronous-provider-for-async-apis

**Comment by Reprevise** [comment]
@rrousselGit Would this still be supported? This is a common pattern as specified in [this article](https://codewithandrea.com/articles/flutter-presentation-layer/).

```dart
@riverpod
class SignInScreenController extends _$SignInScreenController {
  @override
  FutureOr<void> build() {
    // no-op
  }

  Future<void> signInAnonymously() async {
    final authRepository = ref.read(authRepositoryProvider);
    state = const AsyncLoading();
    state = await AsyncValue.guard(() => authRepository.signInAnonymously());
  }
}
```

**Comment by rrousselGit** [comment]
No. This is the very thing that we're looking to discourage. You're not supposed to do that.

**Comment by Reprevise** [comment]
> No. This is the very thing that we're looking to discourage. You're not supposed to do that.

So currently what is the best way to do it then? Or does Riverpod just not support view controllers in that sense?

**Comment by Reprevise** [comment]
> How to do what exactly?

This is currently the pattern that is used:
![image](https://github.com/rrousselGit/riverpod/assets/25782944/41bf59dc-3d42-4c88-9b40-be9aebb9a597)

If I wanted to do something like MVC with Riverpod, is that just not advised? Because using a no-op `build` method seems like the only way to accomplish this using the Notifier syntax. The only other way I can think of is to use ChangeNotifier, which doesn't even have riverpod_generator support.

I want to have a controller for each of my "view" widgets so that way the widget isn't directly calling a repository.

**Comment by Reprevise** [comment]
> I don't see the relationship between the graph shown and an empty build method.

The `SignInController` is an AsyncNotifier<void> with a no-op build method. Since that is not advisable, the only other way that I think it can be done is via a ChangeNotifier, which has no riverpod_generator support.

**Comment by AhmedLSayed9** [comment]
@Reprevise 
This should be covered by https://github.com/rrousselGit/riverpod/issues/1660.

Currently, I'm using `Option` from fpDart and use `None` to indicate this is an `initial` state. check [this](https://github.com/AhmedLSayed9/deliverzler/blob/master/lib/auth/presentation/providers/sign_in_provider.dart)

You can also use some custom state instead of `Option` or an enum if your state has no data to be preserved in the state, i.e:
```dart
enum SignInState {
  idle,
  success,
}

@riverpod
class SignInScreenController extends _$SignInScreenController {
  @override
  FutureOr<SignInState> build()=> SignInState.idle;

  Future<void> signInAnonymously() async {
    final authRepository = ref.read(authRepositoryProvider);
    state = const AsyncLoading();
    state = await AsyncValue.guard(() => authRepository.signInAnonymously().then((_) => SignInState.success););
  }
}
```

**Comment by rrousselGit** [comment]
> > I don't see the relationship between the graph shown and an empty build method.
> 
> The `SignInController` is an AsyncNotifier with a no-op build method. Since that is not advisable, the only other way that I think it can be done is via a ChangeNotifier, which has no riverpod_generator support.

Why?

I don't have a single clue about why `SignInController` has to have an empty build method or what this has to do with MVC or which problem it's trying to solve.

Please be more specific about the problem at hand. I do not understand at all what we're talking about here 😛

**Comment by neiljaywarner** [comment]
@bizz84 @rrousselGit

**Comment by neiljaywarner** [comment]
@bizz84 @rrousselGit  i know you're both working on it but i would especially love to see (or help build or see something close to) sort of a drag and drop robust solid pattern for an Async  that a experienced flutter dev can make a sample of in a project for a brand new flutter dev (there usually are many and flutter usually tries to make it easy for them) that "just works"

Given
a) do async code and show snackbar or modal dialog or go to another screen
b) do async code and modify a list or similar
It feels like 20-40% or more of our screens/features fit closer into a pattern a then b, and it would be nice to optimize for that (ie login, save to playlist, possible reply etc, possibly favorite etc)

https://riverpod.dev/docs/essentials/side_effects#going-further-showing-a-spinner--error-handling is great *except*
1) It doesn't *seem* (at least to me, i could be wrong) to fit A from above as well (yes, i know you could make an enum or sealed for Blah.success, Blah.falure, Blah.waiting but isn't that sort of duplicating asyncvalue? unless i'm confused
2) again, i could be wrong but for new to intermediate riverpod devs it feels like a "Selling point" /key point is to generally favor AsyncValue when() over FutureProvider and avoid StatefulWidgets generally.  So far to me useBlahController and useState seems straightforward and simple and quick and easy but useFuture(possiblyuseMemoized) did not at least at first glance, but i did not come from react i just like simple hooks over stateful widget boilerplate

sorry for long comment hopeful more helpful than not

**Comment by rrousselGit** [comment]
@neiljaywarner I think you're just looking for #1660 

The pattern shown in that doc is only "In a world without mutations".

**Comment by neiljaywarner** [comment]
@rrousselGit  thanks so much for your super prompt response, very cool. Yes, I'm following 1660 eagerly but patiently - except, i still do not see how this fits a paradigm like login use case where void seems appropriate bc AsyncError handlesa error and success just means go to next page, there is nothing to "update", there is no "get" yet it is still (I think) a mutation/side effect in the sense that it is POST not GET, not fetching data

**Comment by l-7-l** [comment]
> Hi, new to Riverpod here. This is my case:
> 
> ```dart
> @Riverpod(keepAlive: true)
> class Package extends _$Package {
>   @override
>   PackageInfo? build() {
>     return null;
>   }
> 
>   Future<void> init() async => state = await PackageInfo.fromPlatform();
> }
> ```
> 
> Why? Because I want to directly access the value rather than working with AsyncValue. Do you know a better way?

same here.  ***maybe it's worth considering allowing it if there are no side effects?***
```dart
/// prodvider.dart
@Riverpod(keepAlive: true)
class ProductsNotifier extends _$ProductsNotifier {
  @override
  ProductsState build() => ProductsState.empty();

  Future<void> fetchOneProduct(ID id) async {
    state = state.copyWith(status: const AsyncStatus.loading());

    final res = await ref.read(productsRepositoryProvider).fetchOneProduct(id);

    state = state.copyWith(
      status: const AsyncStatus.success(),
      paramKeys: state.paramKeys.mergeList(res.paramKeys),
      paramValues: state.paramValues.mergeList(res.paramValues),
      products: state.products.add(res.product),
    );
  }
}
```
```dart
/// state.dart
@freezed
class ProductsState with _$ProductsState {
  const factory ProductsState({
    required Normalized<Product> products,
    required Normalized<ParamKey> paramKeys,
    required Normalized<ParamValue> paramValues,
    required IMap<ID, PricingSchema> pricingSchemas,
    required AsyncStatus status,
  }) = _ProductsState;

  factory ProductsState.empty() => ProductsState(
        products: Normalized.empty(),
        paramKeys: Normalized.empty(),
        paramValues: Normalized.empty(),
        pricingSchemas: const IMapConst({}),
        status: const AsyncStatus.idle(),
      );
}
```
```dart
/// call it before page navigation or somethings like useEffect();
onTap: () {
    ref.read(productsNotifierProvider.notifier).fetchOneProduct(id);
    AutoRouter.of(context).push(ProductRoute(cover: cover, id: id));
},
```

**Comment by festelo** [comment]
> Yes, they should always do so.
There are docs which showcase what I would recommend: https://riverpod.dev/docs/essentials/side_effects#going-further-showing-a-spinner--error-handling

Hi! But what if we don't want `build()` to be called?

```dart
@Riverpod()
class UserNotifier extends _$UserNotifier {
  @override
  Future<User> build() {
    print('making request to get user...');
    return Future.value(User());
  }

  Future<void> setUser(User user) async {
    print('making request to update user...');
  }

  Future<void> deleteUser() async {
    print('making request to delete user...');
  }
}
```
Then, if we would call `ref.read(userNotifierProvider.notifier).deleteUser()` somewhere in the code, it would end up in calling `build` first. So it's one more redundant API request. Are there any ways to avoid it except splitting the provider and keeping empty `build`?

<details>
  <summary>More code</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'main.g.dart';

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class User {}

@Riverpod()
class UserNotifier extends _$UserNotifier {
  @override
  Future<User> build() {
    print('making request to get user...');
    return Future.value(User());
  }

  Future<void> setUser(User user) async {
    print('making request to update user...');
  }

  Future<void> deleteUser() async {
    print('making request to delete user...');
  }
}

class MyApp extends ConsumerWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: Center(
        child: TextButton(
          child: const Text('Delete user'),
          onPressed: () => ref.read(userNotifierProvider.notifier).deleteUser(),
        ),
      ),
    );
  }
}
```
-----------
When the button pressed:
```
flutter: making request to get user...
flutter: making request to delete user...
```

</details>

This can be the case if we, for example, want to have user deletion placed deep in settings, on a page that doesn't have any calls to `ref.watch(userNotifierProvider)`.

**Comment by rrousselGit** [comment]
It sounds niche to have a setting page that enables you to delete a user, without first showing you what that user is.

Such a page would typically show the user name/age/...
It might even ask you to type the full First name + Last name before deleting the user, for confirmation.

**Comment by festelo** [comment]
Yes, could be niche, but even in [this](https://riverpod.dev/docs/essentials/side_effects) example there's a button that does an action, but there's no any other widget that reads the state. So everytime .read() is called, unnecesary execution of build() would happen as well.  

> Only a single network-request is needed.
But actually there would be 2 requests, if there would be no other widgets which would initialize `TodoList` beforehand 🙃

With separate `init()` you can control it on your own.

**Comment by rrousselGit** [comment]
That's fairly out of the scope of this issue. If you want that obtaining `.notifier` does not call `Notifier.build`, this is a new feature.

**Comment by ldesordi** [comment]
Some people implement the provider with an empty constructor method, especially in large projects, so that they can implement a provider for each mutation method. Otherwise, implementing the constructor, getter, creater, updater, and delete methods in a single riverpod provider makes the UI very susceptible to losing its state and consequently executing the fetch method before the mutation method, which messes up state management. This is, I think, because riverpod is very invasive and lacks documentation to explain how things work. Despite this, riverpod is the most promising tool we have.

**Comment by ldesordi** [comment]
Some people implement the provider with an empty constructor method, especially in large projects, so that they can implement a provider for each mutation method. This way is possible to implement reusable components, in a way that we have, for exemple, one single DeleteButton for the entire app. Which also makes the app easier to maintain. Otherwise, implementing the build, fetch, create, update, and delete methods in a single riverpod provider, considering that riverpod is very invasive and lacks documentation to explain how things work, makes the UI very susceptible to losing its state and consequently executing the fetch method before the mutation method, which messes up state management. Despite this, riverpod is the most promising tool we have.

**Comment by rrousselGit** [comment]
This is low value. Closing
